stl/inc/xutility

```diff
@@ -2101,13 +2101,12 @@ namespace ranges {
 
             template <class _Ty>
             _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
-                constexpr bool _Is_lvalue = is_lvalue_reference_v<_Ty>;
-
-                if constexpr (is_array_v<remove_reference_t<_Ty>>) {
-                    if constexpr (_Is_lvalue) {
-                        return {_St::_Array, true};
-                    }
-                } else if constexpr (_Is_lvalue && _Has_member<_Ty>) {
+                constexpr bool _Is_rvalue = is_rvalue_reference_v<_Ty>;

```

Comment:
I think I found a bug there. `is_rvalue_reference_v` won't tell you if _Ty is an rvalue, because the `&&` was "eaten" before. It's quite hard to explain so I will give an example:
```
#include <iostream>
using namespace std;

template <typename T>
void f(T&& t) {
    if (is_lvalue_reference_v<T>)
        cout << "lvalue reference\n";
    else if (is_rvalue_reference_v<T>)
        cout << "rvalue reference\n";
    else
        cout << "non-reference type\n";
}

int main() {
    f(int()); // int() is rvalue; prints "non-reference type"
    int i;
    f(i); // i is lvalue; prints "lvalue reference"
    f(move(i)); // move(i) is rvalue reference; prints "non-reference type"

    cin.get();
}
```
So `is_rvalue_reference_v<_Ty>` should be replaced with `!is_lvalue_reference_v<_Ty>`.
This also resolves our conversation about `safe_range` concept.

Reply:
I think you have to use the forwarding reference here:
```cpp
template <typename T>
void f(T&& t) {
    if (is_lvalue_reference_v<T&&>)
        cout << "lvalue reference\n";
    else if (is_rvalue_reference_v<T&&>)
        cout << "rvalue reference\n";
    else
        cout << "non-reference type\n";
}
```

At least that is working as expected

Reply:
Also this is working as intended as the template argument type is explicitely set, whereas in your example it is deduced.
```cpp
template <typename T>
void f() {
    if constexpr(std::is_lvalue_reference_v<T>)
        std::cout << "lvalue reference\n";
    else if constexpr(std::is_rvalue_reference_v<T>)
        std::cout << "rvalue reference\n";
    else
        std::cout << "non-reference type\n";
}

int main() {
    f<int&&>(); 
    f<int&>();
    f<int>();
}
```

Reply:
Yes, so you have to either use `is_rvalue_reference_v<_Ty&&>` or `!is_lvalue_reference_v<_Ty>` (or `!is_lvalue_reference_v<_Ty&&>` if you want). They should produce exactly the same output. I think `!is_lvalue_reference_v<_Ty>` is better because it's more consistent with `safe_range` concept.

Reply:
I mean the whole point of this example was to show what i named "reference eating" and that _Ty will either be lvalue reference or non-reference type but never rvalue referene.

Reply:
But that is exactly how that function is used `requires (_Choice<_Ty>._Strategy != _St::_None)` So consequently there will be a difference
```cpp
static_assert(!is_lvalue_reference_v<int>, "");
static_assert(!is_rvalue_reference_v<int>, "");
```

So one cannot simply replace `is_rvalue_reference_v` with `!is_lvalue_reference_v` without changing behavior. The question is what is the intended behavior.

Reply:
I meant that no one will probably explicitly specify _Ty as rvalue reference in operator()

Reply:
All I mean is that if an rvalue is passed to operator(), _Ty will be a non-reference type and if lvalue is passed _Ty will be lvalue reference. _Ty is then passed to _Choice_t which passes _Ty to _Choose and that is_rvalue_reference_v<_Ty> won't tell if _Ty is an rvalue, but is_lvalue_reference_v<_Ty> can tell if _Ty is an lvalue, so !is_lvalue_reference_v<_Ty> will tell if _Ty is not lvalue and if it's not lvalue it is rvalue.

Reply:
My explanation is probably very hard to understand and sorry about that but it's not easy to explain.

Reply:
Resolved by using the intermediate concept